dicoogle/src/main/java/pt/ua/dicoogle/taskManager/RunningIndexTasks.java

```diff
@@ -54,13 +56,15 @@ public static RunningIndexTasks getInstance() {
 	}
 
 	public RunningIndexTasks() {
-		taskRunningList = new LinkedHashMap<>();
+		taskRunningList = new ConcurrentHashMap<>(SOFT_MAX_RUNNINGTASKS, 0.75f, 4);
 	}
 
 	public void addTask(String taskUid, Task<Report> task) {
 		taskRunningList.put(taskUid, task);
-		if (ENABLE_HOOK){
+		if (ENABLE_HOOK && !cleaning){
+		    cleaning = true; // will execute cleaning process
 		    hookRemoveRunningTasks();
+		    cleaning = false; // already cleaned
```

Comment:
This is definitely not thread safe.

- Changes to `cleaning` are not guaranteed to be visible to other threads (not `volatile`)
- Even if it was `volatile`, the threads could in theory check the condition `!cleaning` at the same time before setting it to `true`, and both enter the cleaning routine.

This can become thread safe by making `cleaning` an `AtomicBoolean` and changing the code to this:
 
```suggestion
		if (ENABLE_HOOK && !cleaning.compareAndSet(false, true)){
		    // will execute cleaning process
		    hookRemoveRunningTasks();
		    cleaning.set(false); // already cleaned
```